container/lxd/lxd.go

```diff
@@ -95,74 +135,278 @@ func (manager *containerManager) Namespace() instance.Namespace {
+	}
+	return errors.Trace(manager.removeInstance(string(id)))
+}
+
 func (manager *containerManager) CreateContainer(
 	instanceConfig *instancecfg.InstanceConfig,
 	cons constraints.Value,
 	series string,
 	networkConfig *container.NetworkConfig,
 	storageConfig *container.StorageConfig,
 	callback environs.StatusCallbackFunc,
-) (inst instance.Instance, hc *instance.HardwareCharacteristics, err error) {
+) (instance.Instance, *instance.HardwareCharacteristics, error) {
+	callback(status.Provisioning, "Creating container", nil)
+	name, err := manager.createInstance(instanceConfig, cons, series, networkConfig, storageConfig, callback)
+	if err != nil {
+		callback(status.ProvisioningError, fmt.Sprintf("Creating container: %v", err), nil)
+		return nil, nil, errors.Trace(err)
+	}
 
-	defer func() {
-		if err != nil {
-			callback(status.ProvisioningError, fmt.Sprintf("Creating container: %v", err), nil)
+	callback(status.Provisioning, "Container created, starting", nil)
+	err = manager.startInstance(name)
+	if err != nil {
+		if err := manager.removeInstance(name); err != nil {
+			logger.Errorf("Cannot remove failed instance: %s", err)
 		}
-	}()
+		callback(status.ProvisioningError, fmt.Sprintf("Starting container: %v", err), nil)
+		return nil, nil, err
+	}
 
-	if manager.client == nil {
-		manager.client, err = ConnectLocal()
+	callback(status.Running, "Container started", nil)
+	return &lxdInstance{name, manager.server}, &instance.HardwareCharacteristics{AvailabilityZone: &manager.availabilityZone}, nil
+}
+
+func (manager *containerManager) ListContainers() (result []instance.Instance, err error) {
+	result = []instance.Instance{}
+	if manager.server == nil {
+		manager.server, err = ConnectLocal()
 		if err != nil {
-			err = errors.Annotatef(err, "failed to connect to local LXD")
 			return
 		}
 	}
+	lxdInstances, err := manager.server.GetContainers()
+
+	if err != nil {
+		return
+	}
+
+	for _, i := range lxdInstances {
+		if strings.HasPrefix(i.Name, manager.namespace.Prefix()) {
+			result = append(result, &lxdInstance{i.Name, manager.server})
+		}
+	}
+
+	return result, nil
+}
+
+func (manager *containerManager) IsInitialized() bool {
+	if manager.server != nil {
+		return true
+	}
+
+	// NewClient does a roundtrip to the server to make sure it understands
+	// the versions, so all we need to do is connect above and we're done.
+	var err error
+	manager.server, err = ConnectLocal()
+	return err == nil
+}
 
-	// It is only possible to provision LXD containers
-	// of the same architecture as the host.
-	hostArch := arch.HostArch()
+func lxdSocketPath() string {
+	// LXD socket is different depending on installation method
+	// We prefer upstream's preference of snap installed LXD
+	debianSocket := filepath.FromSlash("/var/lib/lxd")
+	snapSocket := filepath.FromSlash("/var/snap/lxd/common/lxd")
+	path := os.Getenv("LXD_DIR")
+	if path != "" {
+		logger.Debugf("Using environment LXD_DIR as socket path: %q", path)
```

Comment:
We're assuming that the env var is correct?  Not verifying it?

Reply:
We kindly respect operators freedom to shoot himself in the foot.